### ISSUE TITLE
Implement OAuth for remote accounts

### ISSUE THREAD

**Issue by exactly-one-kas** [Murmele/Gittyup#226]
This would streamline the login process for GitHub and not require the user to create a personal access token

`QOAuth2AuthorizationCodeFlow` would probably be the best starting point for this

**Comment by Arxareon** [comment]
Plus one for this one!
Many other clients support OAuth login and it's a lot easier and faster - not only for GitHub but others like Bitbucket as well.

An implementation should definitely be considered, and if and when it does, could we get an ETA?

Please and thank you!

**Comment by AHSauge** [comment]
I'm contemplating digging into OAuth with PKCE in another context, and I'm quite willing to feed this in here as well. When I briefly looked at it before, I concluded that this should be doable for GitHub with OAuth with a device workflow set. Similarly it should be doable for GitLab and Gitea as well. This would eliminate the need for a client secret, which there is some Gittyup code eluding to needed in the past.

@Murmele If I provide a PR with the necessary framework to get OAuth working on for example GitLab and GitHub, would you be able to register for a client ID for that? I assume it's best if this is managed along with the source code at least.

**Comment by Murmele** [comment]
> I'm contemplating digging into OAuth with PKCE in another context, and I'm quite willing to feed this in here as well. When I briefly looked at it before, I concluded that this should be doable for GitHub with OAuth with a device workflow set. Similarly it should be doable for GitLab and Gitea as well. This would eliminate the need for a client secret, which there is some Gittyup code eluding to needed in the past.
> 
> [@Murmele](https://github.com/Murmele) If I provide a PR with the necessary framework to get OAuth working on for example GitLab and GitHub, would you be able to register for a client ID for that? I assume it's best if this is managed along with the source code at least.

You mean for testing or as testcase?

**Comment by AHSauge** [comment]
> > I'm contemplating digging into OAuth with PKCE in another context, and I'm quite willing to feed this in here as well. When I briefly looked at it before, I concluded that this should be doable for GitHub with OAuth with a device workflow set. Similarly it should be doable for GitLab and Gitea as well. This would eliminate the need for a client secret, which there is some Gittyup code eluding to needed in the past.
> > [@Murmele](https://github.com/Murmele) If I provide a PR with the necessary framework to get OAuth working on for example GitLab and GitHub, would you be able to register for a client ID for that? I assume it's best if this is managed along with the source code at least.
> 
> You mean for testing or as testcase?

No, in order to bring it into Gittyup and make use of it there.

To elaborate a bit on this, in order to do OAuth you need to register the application with the entity that you want to log into, for example GitHub. Doing so, you will get a `Client ID` and a `Client Secret` that the application must use to utilise OAuth. This is not per user, but per application. For a website or similar, this registration enables the OAuth server to redirect the user back to the website the user logs into. In all cases, the registration provides the identifying details the user sees when interacting with OAuth login or managing OAuth app. access to their user. In the case of GitHub, you can also press a button to list it in GitHub marketplace.

There are some traces of this in the Gittyup code base. However, the `Client Secret` part is somewhat obsolete and insecure in context of a distributed application. This is where the PKCE comes in (in GitHub, named Device Flow). I didn't look into the details, but essentially, you embed the publicly known `Client ID` into the application (source code in this case), and then do essentially a handshake with the OAuth server to log in. From my POV, I can contribute code into Gittyup to handle OAuth, but if I also contribute a `Client ID`, I get (at least in the case of GitHub) access to a button to revoke all tokens associated with that client. I think that just forces everyone to log in again, but as you can imagine, you might want to control this one yourself 😉 

Hence me asking if I contribute code, would you be willing to do the necessary registration for it?
In the case of Gittyup, this is tied to your user, so for GitHub you register this in https://github.com/settings/developers
